File: tests/snapshots/sum_types.c
// Data-carrying enums (sum types): each variant carries its own data. Build one
// with `Shape.Circle(r)`; pull the data back out by matching. This is the same
// shape as Option/Result — now you can define your own.

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

typedef enum { Shape_Circle, Shape_Rectangle, Shape_Empty } ShapeTag;
typedef struct {
    ShapeTag tag;
    union {
        struct { double f0; } Circle;
        struct { double f0; double f1; } Rectangle;
    } data;
} Shape;

typedef struct { bool is_ok; double ok; char* err; } Result_double_str;
static double Result_double_str_unwrap(Result_double_str r) { if (!r.is_ok) { fprintf(stderr, "unwrapped an Err\n"); exit(1); } return r.ok; }

static char* vbr_dup(const char* s) {
    char* d = (char*)malloc(strlen(s) + 1);
    strcpy(d, s);
    return d;
}
static char* vbr_from_double(double d) {
    char buf[64];
    for (int p = 1; p <= 17; p++) {
        snprintf(buf, sizeof buf, "%.*g", p, d);
        if (strtod(buf, NULL) == d) break;
    }
    return vbr_dup(buf);
}
static char* vbr_concat(const char* a, const char* b) {
    char* s = (char*)malloc(strlen(a) + strlen(b) + 1);
    strcpy(s, a);
    strcat(s, b);
    return s;
}

Result_double_str area(Shape s);

Result_double_str area(Shape s) {
    Shape _m0 = s;
    if (_m0.tag == Shape_Circle) {
        double r = _m0.data.Circle.f0;
        return (Result_double_str){ .is_ok = true, .ok = ((3.14159 * r) * r) };
    } else if (_m0.tag == Shape_Rectangle) {
        double w = _m0.data.Rectangle.f0;
        double h = _m0.data.Rectangle.f1;
        return (Result_double_str){ .is_ok = true, .ok = (w * h) };
    } else {
        return (Result_double_str){ .is_ok = true, .ok = 0.0 };
    }
}

int main(void) {
    Shape c = (Shape){ .tag = Shape_Circle, .data.Circle = { 2.0 } };
    Shape r = (Shape){ .tag = Shape_Rectangle, .data.Rectangle = { 3.0, 4.0 } };
    Result_double_str _t0 = area(c);
    if (!_t0.is_ok) { fprintf(stderr, "Error: %s\n", _t0.err); return 1; }
    printf("%s\n", vbr_concat("circle area = ", vbr_from_double(_t0.ok)));
    Result_double_str _t1 = area(r);
    if (!_t1.is_ok) { fprintf(stderr, "Error: %s\n", _t1.err); return 1; }
    printf("%s\n", vbr_concat("rect area   = ", vbr_from_double(_t1.ok)));
    Result_double_str _t2 = area((Shape){ .tag = Shape_Empty });
    if (!_t2.is_ok) { fprintf(stderr, "Error: %s\n", _t2.err); return 1; }
    printf("%s\n", vbr_concat("empty area  = ", vbr_from_double(_t2.ok)));
    return 0;
}

File: tests/snapshots/sum_types.py
# Data-carrying enums (sum types): each variant carries its own data. Build one
# with `Shape.Circle(r)`; pull the data back out by matching. This is the same
# shape as Option/Result — now you can define your own.

import sys
from dataclasses import dataclass

@dataclass
class Ok:
    value: object

@dataclass
class Err:
    error: object

def _vb(x):
    if isinstance(x, bool):
        return "true" if x else "false"
    if isinstance(x, float) and x.is_integer():
        return str(int(x))
    return str(x)

class Shape:
    pass

@dataclass
class Circle(Shape):
    f0: float

@dataclass
class Rectangle(Shape):
    f0: float
    f1: float

@dataclass
class Empty(Shape):
    pass

def area(s: Shape) -> float:
    _m0 = s
    match _m0:
        case Circle(r):
            return Ok((3.14159 * r) * r)
        case Rectangle(w, h):
            return Ok(w * h)
        case Empty():
            return Ok(0.0)

def main():
    c: Shape = Circle(2.0)
    r: Shape = Rectangle(3.0, 4.0)
    _t0 = area(c)
    if isinstance(_t0, Err):
        print(f"Error: {_t0.error}", file=sys.stderr)
        raise SystemExit(1)
    print(f"circle area = {_vb(_t0.value)}")
    _t1 = area(r)
    if isinstance(_t1, Err):
        print(f"Error: {_t1.error}", file=sys.stderr)
        raise SystemExit(1)
    print(f"rect area   = {_vb(_t1.value)}")
    _t2 = area(Empty())
    if isinstance(_t2, Err):
        print(f"Error: {_t2.error}", file=sys.stderr)
        raise SystemExit(1)
    print(f"empty area  = {_vb(_t2.value)}")


if __name__ == "__main__":
    main()
